File: matlab/figure4.m
% FIGURE4.M
% Equation-by-equation estimates with bootstrap confidence intervals

clear;

trivar

%%%%%%%%%% Historical decomposition %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Compute structural multipliers
IRF=irfvar(A,SIGMA(1:q,1:q),p,t-p-1);

% Compute structural shocks Ehat from reduced form shocks Uhat
Ehat=inv(chol(SIGMA)')*Uhat(1:q,:);

% Cross-multiply the weights for the effect of a given shock on the real
% oil price (given by the relevant row of IRF) with the structural shock
% in question
yhat1=zeros(t-p,1); yhat2=zeros(t-p,1); yhat3=zeros(t-p,1);
for i=1:t-p
	yhat1(i,:)=dot(IRF(3,1:i),Ehat(1,i:-1:1));
	yhat2(i,:)=dot(IRF(6,1:i),Ehat(2,i:-1:1));
	yhat3(i,:)=dot(IRF(9,1:i),Ehat(3,i:-1:1));
end;

% Contribution to real price of oil
time=1975+2/12:1/12:2007+12/12;

subplot(3,1,1)
plot(time,yhat1,'b-');
title('Cumulative Effect of Oil Supply Shock on Real Price of Crude Oil')
axis([1976+1/12 2007+12/12 -100 +100])
grid on

subplot(3,1,2)
plot(time,yhat2,'b-');
title('Cumulative Effect of Aggregate Demand Shock on Real Price of Crude Oil')
axis([1976+1/12 2007+12/12 -100 +100])
grid on

subplot(3,1,3)
plot(time,yhat3,'b-');
title('Cumulative Effect of Oil-Market Specific Demand Shock on Real Price of Crude Oil')
axis([1976+1/12 2007+12/12 -100 +100])
grid on

File: figure4.R
# 必要なライブラリの読み込み
library(vars)
library(ggplot2)
library(grid)
library(gridExtra)

# 別ファイルに定義した関数の読み込み
source("modules/trivar.R")   # inputは行列データ、outputはvar_model, Uhat, Ehat, SIGMA, time, h, p
source("modules/make_matrix.R")   # inputはファイルパス、outputは行列化されたデータ

# 読み込んだ関数の実行
data_matrix <- make_matrix("data/data.txt")
trivar_res <- trivar(data_matrix)

# varsパッケージのirf関数で、インパルス応答関数を得る
IRF <- irf(
  trivar_res$var_model,
  impulse = colnames(trivar_res$Uhat),
  response = colnames(trivar_res$Uhat),
  boot = FALSE,
  n.ahead = length(trivar_res$time) - trivar_res$p - 1
)

# 構造的ショックEhatを計算
Ehat <- solve(t(chol(trivar_res$SIGMA))) %*% t(trivar_res$Uhat)

# 応答変数の予測値をそれぞれ時系列順に格納
yhat1 <- yhat2 <- yhat3 <- rep(0, length(trivar_res$time) - trivar_res$p)
for(i in 1:(length(trivar_res$time) - trivar_res$p)){
  yhat1[i] <- sum(IRF$irf$V1[1:i,1] * Ehat[1, i:1])  
  yhat2[i] <- sum(IRF$irf$V2[1:i,2] * Ehat[2, i:1])  
  yhat3[i] <- sum(IRF$irf$V3[1:i,3] * Ehat[3, i:1])  
}

# データをプロット
# プロットするためにデータフレームを作成
plot_data1 <- data.frame(
  time = trivar_res$time[(trivar_res$p + 1):length(trivar_res$time)],
  yhat = yhat1
)
plot_data2 <- data.frame(
  time = trivar_res$time[(trivar_res$p + 1):length(trivar_res$time)],
  yhat = yhat2
)
plot_data3 <- data.frame(
  time = trivar_res$time[(trivar_res$p + 1):length(trivar_res$time)],
  yhat = yhat3
)

# プロットのロジックを関数化
figure4_plotter <- function(plot_data, yhat, subtitle) {
  result <- ggplot(plot_data, aes(x = time)) +
    geom_path(aes(y = yhat), color = 'blue', linewidth = 1) +
    ylim(-100, 100) +
    labs(title = subtitle,
         y = "",                   
         x = "") +
    theme(plot.title = element_text(hjust = 0.5))
  return(result)
}

# 定義した関数を実行
pl1 <- figure4_plotter(plot_data1, yhat1, "Cumulative Effect of Oil Supply Shock on Real Price of Crude Oil")
pl2 <- figure4_plotter(plot_data2, yhat2, "Cumulative Effect of Aggregate Demand Shock on Real Price of Crude Oil")
pl3 <- figure4_plotter(plot_data3, yhat3, "Cumulative Effect of Oil-Market Specific Demand Shock on Real Price of Crude Oil")

# 縦一列にプロット
gridExtra::grid.arrange(
  pl1, 
  pl2, 
  pl3, 
  nrow = 3, 
  top = textGrob(
    expression(
      bold("Historical Decomposition of Real Price of Oil 1976.1-2007.12")),
    gp = gpar(col = "black", fontsize = 20)
  )
)
